src/hotspot/share/compiler/precompiler.cpp

```diff
@@ -232,16 +234,24 @@ void Precompiler::compile_cached_code(ArchiveBuilder* builder, TRAPS) {
       pi.precompile(builder, THREAD);
     }
 
-    for (int level = CompLevel_simple; level <= CompLevel_full_optimization; level++) {
-      CompLevel comp_level = (CompLevel)level;
-      if (comp_level == CompLevel_full_profile) {
-        comp_level = CompLevel_limited_profile;
-      }
-      PrecompileIterator pi(comp_level, false /*for_preload*/, (CompLevel)level, THREAD);
+    {
+      PrecompileIterator pi(CompLevel_full_optimization, false /*for_preload*/, CompLevel_full_optimization, THREAD);
       TrainingData::iterate(pi);
       pi.precompile(builder, THREAD);
     }
 
+    if (TieredCompilation) {
```

Comment:
How does this work as far as consistency between training/assembly run and production run is concerned? Do we need to enforce the same setting across these runs or can we e.g. train and the (assuming we fork) assemble with -TieredCompilation but production run with +TieredCompilation -- or vice versa?

Reply:
I tested different combinations of `{+|-}TieredCompilation` with training and production runs, and it works well. I believe the normal mechanisms make it right. No C1 code? Nothing loaded. C1 code in archive, but C1 is disabled? We are not loading it then. 

In fact, there seem to be a slight performance edge in training with C2 only, and then running with C1+C2: no penalty on loading lots of C1 compiled code, and quicker transition from intepreter to "real" C1 in only a few places where it is needed! Compiler dynamics is weird.

Reply:
As I mentioned during the meeting, it looks weird to me to see `if (TieredCompilation)` checks in Precompiler code which is executed only during assembly phase. As a data point, regular CDS archive is used to be generated in interpreter-only mode (to make the process reproducible). 

Instead of supporting arbitrary JVM configurations during assembly phase, I'd prefer to see offending JVM parameters being overridden/dropped.

Reply:
Right, I thought it matters for Precompiler not to attempt pre-compiling C1 classes with `-XX:-TieredCompilation`, but it does not. I see `runtime/cds` passing with/without TieredCompilation without these hunks, and CDS archive sizes are similar, and performance is still the same. So I have reverted Precompiler hunks as unnecessary.

Take a look again?

Reply:
> CDS archive sizes are similar,

I think this would impact the code archive, not the CDS archive. 
If it does reduce the on-disk size of code archive, then why not do this? It's pretty straightforward; not a complicated change IMO.

Reply:
Do you see the same kind of improvement with new version of the patch and `-XX:-TieredCompilation`?

Reply:
> Do you see the same kind of improvement with new version of the patch and -XX:-TieredCompilation?

aah! I get it now. For some reason I kept thinking we are not able to run with `-XX:-TieredCompilation` in the training/assembly run and this particular hunk fixes it. But that's not the case. My bad!

Reply:
No problem! The interactions between different phases and compiler tiers are often confusing :)